packages/flutter/test/widgets/navigator_replacement_test.dart

```diff
@@ -39,42 +42,205 @@ void main() {
-
-    navigator.pushNamedAndRemoveUntil('/a', ModalRoute.withName('/b'));
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsNothing);
-    expect(find.text('a', skipOffstage: false), findsOneWidget);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
+  group('pushAndRemoveUntil', () {
+
+    testWidgets('notifies appropriately', (WidgetTester tester) async {
+      final TestObserver observer = TestObserver();
+      final Widget myApp = MaterialApp(
+        home: const Material(child: Text('home')),
+        routes: <String, WidgetBuilder>{
+          '/a': (BuildContext context) => const Material(child: Text('a')),
+          '/b': (BuildContext context) => const Material(child: Text('b')),
+        },
+        navigatorObservers: <NavigatorObserver>[observer],
+      );
+
+      await tester.pumpWidget(myApp);
+
+      final NavigatorState navigator = tester.state(find.byType(Navigator));
+      navigator.pushNamed('/a');
+      await tester.pumpAndSettle();
+
+      expect(find.text('home', skipOffstage: false), findsOneWidget);
+      expect(find.text('a', skipOffstage: false), findsOneWidget);
+      expect(find.text('b', skipOffstage: false), findsNothing);
+
+      final List<String> log = <String>[];
+      observer
+        ..onPushed = (Route<dynamic> route, Route<dynamic> previousRoute) {
+          expectSync(route is PageRoute && route.settings.name == '/b', isTrue);
+          expectSync(previousRoute is PageRoute && previousRoute.settings.name == '/a', isTrue);
+          log.add('/b pushed, previous route: /a');
+        }
+        ..onRemoved = (Route<dynamic> route, Route<dynamic> previousRoute) {
+          if (route.settings.name == '/a') {
+            expectSync(previousRoute, null);
+            log.add('/a removed, previous route: null');
+          } else if (route.settings.name == '/') {
+            expectSync(previousRoute, null);
+            log.add('/ removed, previous route: null');
+          }
+        };
+
+      // Remove all routes below
+      navigator.pushNamedAndRemoveUntil('/b', (Route<dynamic> route) => false);
+      await tester.pumpAndSettle();
+
+      expect(find.text('home', skipOffstage: false), findsNothing);
+      expect(find.text('a', skipOffstage: false), findsNothing);
+      expect(find.text('b', skipOffstage: false), findsOneWidget);
+      expect(log, contains('/b pushed, previous route: /a'));
+      expect(log, contains('/a removed, previous route: null'));
+      expect(log, contains('/ removed, previous route: null'));
+
+      observer.onPushed = null;
+      log.clear();
+
+
+      navigator.pushNamed('/');
+      await tester.pumpAndSettle();
+
+      expect(find.text('home', skipOffstage: false), findsOneWidget);
+      expect(find.text('a', skipOffstage: false), findsNothing);
+      expect(find.text('b', skipOffstage: false), findsOneWidget);
+
+      observer
```

Comment:
If we do the callback registration generically as suggested above, we don't have to repeat it here...